libraries/AC_Avoidance/AP_OADijkstra.cpp

```diff
@@ -218,103 +260,321 @@ bool AP_OADijkstra::create_polygon_fence_with_margin(float margin_cm)
+    // clear all points
+    _exclusion_polygon_numpoints = 0;
+
+    // return immediately if no exclusion polygons
+    const uint16_t num_exclusion_polygons = fence->polyfence().get_exclusion_polygon_count();
+
+    // iterate through exclusion polygons and create outer points
+    for (uint16_t i = 0; i < num_exclusion_polygons; i++) {
+        uint16_t num_points;
+        const Vector2f* boundary = fence->polyfence().get_exclusion_polygon(i, num_points);
+        if (num_points < 3) {
+            // ignore exclusion polygons with less than 3 points
+            continue;
+        }
+
+        // expand array if required
+        if (!_exclusion_polygon_pts.expand_to_hold(_exclusion_polygon_numpoints + num_points)) {
+            return false;
+        }
+
+        // for each point in exclusion polygon
+        // Note: boundary is "unclosed" meaning the last point is *not* the same as the first
+        for (uint8_t j=0; j<num_points; j++) {
+
+            // find points before and after current point (relative to current point)
+            const uint8_t before_idx = (j == 0) ? num_points-1 : j-1;
+            const uint8_t after_idx = (j == num_points-1) ? 0 : j+1;
+            Vector2f before_pt = boundary[before_idx] - boundary[j];
+            Vector2f after_pt = boundary[after_idx] - boundary[j];
+
+            // if points are overlapping fail
+            if (before_pt.is_zero() || after_pt.is_zero() || (before_pt == after_pt)) {
+                return false;
+            }
+
+            // scale points to be unit vectors
+            before_pt.normalize();
+            after_pt.normalize();
+
+            // calculate intermediate point and scale to margin
+            Vector2f intermediate_pt = (after_pt + before_pt) * 0.5f;
+            float intermediate_len = intermediate_pt.length();
+            intermediate_pt *= (margin_cm / intermediate_len);
+
+            // find final point which is outside the original polygon
+            uint16_t next_index = _exclusion_polygon_numpoints + j;
+            _exclusion_polygon_pts[next_index] = boundary[j] + intermediate_pt;
+            if (!Polygon_outside(_exclusion_polygon_pts[next_index], boundary, num_points)) {
+                _exclusion_polygon_pts[next_index] = boundary[j] - intermediate_pt;
+                if (!Polygon_outside(_exclusion_polygon_pts[next_index], boundary, num_points)) {
+                    // could not find a point on either side that was outside the exclusion polygon so fail
+                    // this may happen if the exclusion polygon has overlapping lines
+                    return false;
+                }
+            }
+        }
+
+        // update total number of points
+        _exclusion_polygon_numpoints += num_points;
+    }
+
+    // record fence update time so we don't process these exclusion polygons again
+    _exclusion_polygon_update_ms = fence->polyfence().get_exclusion_polygon_update_ms();
```

Comment:
OK, I see your point.  The way Dijkstra's update() function is written it will keep trying to create_exclusion_polygon_with_margin() over and over even if the underlying data hasn't changed.  I've created [this issue ](https://github.com/ArduPilot/ardupilot/issues/12204) to capture this enhancement request.  I think this is an enhancement though so we can leave the change out of this PR